acvm-repo/acir_field/src/field_element.rs

```diff
@@ -0,0 +1,544 @@
+                    mul_sign,
+                    smaller_repr / &power_of_two,
+                );
+            }
+        }
+        write!(f, "{smaller_repr}")
+    }
+}
+
+impl<F: PrimeField> std::fmt::Debug for FieldElement<F> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        std::fmt::Display::fmt(self, f)
+    }
+}
+
+impl<F: PrimeField> std::hash::Hash for FieldElement<F> {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        state.write(&self.to_be_bytes());
+    }
+}
+
+impl<F: PrimeField> PartialEq for FieldElement<F> {
+    fn eq(&self, other: &Self) -> bool {
+        self.to_be_bytes() == other.to_be_bytes()
+    }
+}
+
+impl<F: PrimeField> From<i128> for FieldElement<F> {
+    fn from(mut a: i128) -> FieldElement<F> {
+        let mut negative = false;
+        if a < 0 {
+            a = -a;
+            negative = true;
+        }
+
+        let mut result = match F::from_str(&a.to_string()) {
+            Ok(result) => result,
+            Err(_) => panic!("Cannot convert i128 as a string to a field element"),
+        };
+
+        if negative {
+            result = -result;
+        }
+        FieldElement(result)
+    }
+}
+
+impl<T: ark_ff::PrimeField> Serialize for FieldElement<T> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        self.to_hex().serialize(serializer)
+    }
+}
+
+impl<'de, T: ark_ff::PrimeField> Deserialize<'de> for FieldElement<T> {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let s: Cow<'de, str> = Deserialize::deserialize(deserializer)?;
+        match Self::from_hex(&s) {
+            Some(value) => Ok(value),
+            None => Err(serde::de::Error::custom(format!("Invalid hex for FieldElement: {s}",))),
+        }
+    }
+}
+
+impl<F: PrimeField> From<u128> for FieldElement<F> {
+    fn from(a: u128) -> FieldElement<F> {
+        let result = match F::from_str(&a.to_string()) {
+            Ok(result) => result,
+            Err(_) => panic!("Cannot convert u128 as a string to a field element"),
+        };
+        FieldElement(result)
+    }
+}
+
+impl<F: PrimeField> From<usize> for FieldElement<F> {
+    fn from(a: usize) -> FieldElement<F> {
+        FieldElement::from(a as u128)
+    }
+}
+
+impl<F: PrimeField> From<bool> for FieldElement<F> {
+    fn from(boolean: bool) -> FieldElement<F> {
+        if boolean {
+            FieldElement::one()
+        } else {
+            FieldElement::zero()
+        }
+    }
+}
+
+impl<F: PrimeField> FieldElement<F> {
+    pub fn from_repr(field: F) -> Self {
+        Self(field)
+    }
+
+    // XXX: This method is used while this field element
+    // implementation is not generic.
+    pub fn into_repr(self) -> F {
+        self.0
+    }
+
+    fn is_negative(&self) -> bool {
+        self.neg().num_bits() < self.num_bits()
```

Comment:
as far as I can tell, this will be incorrect when the field element is in the middle of its range. e.g. if we had two bit field elements, `-2` and `2` will both have the same number of bits within the representation.

instead, I believe we should compare without `num_bits`: `self.neg() < self`

Reply:
I'd like to avoid changing this in this PR as it's an actual logic change whereas this PR is mostly just organising existing code.

Reply:
made issue: https://github.com/noir-lang/noir/issues/5174